hjson/hjson_test.go

```diff
@@ -100,17 +88,28 @@ func TestArray(t *testing.T) {
 		c: 4
 	}
 ]
-`)
-
-	c, err := NewConfig(input)
-	require.NoError(t, err, "failed to parse input")
-
+`
+	c := mustNewConfig(t, input)
 	var verify []map[string]int
-	err = c.Unpack(&verify)
-	require.NoError(t, err, "failed to unpack config")
+	mustUnpack(t, c, &verify)
 	require.Len(t, verify, 2)
 
 	assert.Equal(t, verify[0]["b"], 2)
 	assert.Equal(t, verify[0]["c"], 3)
 	assert.Equal(t, verify[1]["c"], 4)
 }
+
+// mustNewConfig asserts that a new configuration object creation from the given Hjson
+// string with or without options was successful and returned no error (i.e. `nil`).
+func mustNewConfig(t *testing.T, input string, opts ...ucfg.Option) *ucfg.Config {
```

Comment:
In case we introduce more of these, how about providing a `ucfgtest` package with these 2 as helpers?

A third helper comming to mind would be `MustBuildAndUnpack(t *testing.t, to interface{}, input string, opts ...ucfg.Option)`

Reply:
Yeah, that would be a good idea. I was on the fence about introducing a new package and moving these helper functions. But later realized that we still need to use the `NewConfig` from `hjson/json/yaml `packages inside `MustNewConfig` even when we move the `MustNewConfig` to `ucfgtest`. I’m wondering what would be your approach to move these helpers into `ucfgtest`. Also the third helper that is in your mind, `MustBuildAndUnpack(t *testing.t, to interface{}, input string, opts ...ucfg.Option)`, is that for combining both configuration object creation and unpack together in a single shot?

Reply:
I was thinking to add the helper package at the top-level (package name `github.com/elastic/go-ucfg/ucfgtest`). The idea is that the helpers would not only be available internally, but can be used by other projects that build on go-ucfg. For sharing code between internal packages only if we would use `github.com/elastic/go-ucfg/internal/<shared>` in order to not "increase" the public interface just for internal helpers. E.g. there are projects that build a configuration dynamically by combining multiple sources. Having `MustUnpack` can be usefull for validation purposes (or even have a `MustMatch`, which compares the structure of 2 configs). The drawback of `ucfgtest` is that we can not use it in the top-level package out of the box. Although we can keep the files in place, we need to adapt the `*_test.go` files to begin with `package ucfg_test`, in order to have an independent package that can import `ucfgteset` (break circular dependencies).

> Also the third helper that is in your mind, MustBuildAndUnpack(t *testing.t, to interface{}, input string, opts ...ucfg.Option), is that for combining both configuration object creation and unpack together in a single shot?

Yes. But thinking about it, I don't think we really need this. Let's keep it out.

Reply:
First, apologies for the delayed reply. I got the idea of moving these two functions into github.com/elastic/go-ucfg/ucfgtest. But I didn't get the whole picture of how we are going to break the circular dependency to make use of the MustUnpack inside the tests. 

> For sharing code between internal packages only if we would use github.com/elastic/go-ucfg/internal/<shared> in order to not "increase" the public interface just for internal helpers. 

Does this mean that we might need to duplicate the code to share between internal package?

Reply:
>`Does this mean that we might need to duplicate the code to share between internal packages?

No. In the top level package we will need to change `package ucfg` to `package ucfg_test` in the `*_test.go` files. This way the test suite is a "special" package that must import `ucfg` itself, as well as other helpers.

From go test tool help:

> Test files that declare a package with the suffix "_test" will be compiled as a
separate package, and then linked and run with the main test binary.

e.g. a quick google search reveals this [blog post](https://splice.com/blog/lesser-known-features-go-test/) (see "black box test package")